### PR TITLE
List modules/custom SQL IDs for better practices

### PR DESCRIPTION
**Pull request by Knindzagxg**
List of Creature ID's that have been used in Azerothcore Modules.

### CODE REVIEW

**Comment by BarbzYHOOL** [comment]
Give your opinion guys

I think it's a good idea but dunno if it's the right place (but where otherwise??)

**Comment by Knindzagxg** [comment]
Maybe make new folder like ingame information

There is a lot of info that members always search, towards some unused ids, morph ids or etc
I think new folder / subfolder would do the job

**Comment by Stoabrogga** [comment]
Perhaps we should also add NPC text IDs or gossip menu IDs. For example concerning MorphSummon I used these:
NPC text IDs: 601072, 601073, 601074
Gossip menu IDs: 61072, 61073, 61074

**Comment by BarbzYHOOL** [comment]
Things to add to the list: creature ID / GUID, gossip menu ID, acore_string, game object ID / GUID, npc text ID, broadcast text ID

Depends on the modules

It needs to be structured correctly (by column imo) and put in the "best practices" for modules

**Comment by Stoabrogga** [comment]
I pushed a commit using a new Markdown document and adding a link to the sidebar. I also included a table containing valid ID ranges as suggestion (the only thing that bugs me there is "gossip_menu" because it uses smallint for MenuID).

**Comment by Stoabrogga** [comment]
Perhaps we could change the gossip "MenuID" to use mediumint instead of smallint, would make all ranges equal and easier to describe (and remember). As far as I saw in the core it is handled via uint16, so should not be a problem. These tables need to be changed:
- gossip_menu
- gossip_menu_option
- gossip_menu_option_locale

**Comment by Stoabrogga** [comment]
Ok, the client is obviously not able to handle gossip menu IDs greater than 65535.

File: docs/Dealing-with-IDs.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# Dealing with IDs
+
+If dealing with IDs it has to be ensured that the same ID is not used twice. Please use these ranges for new IDs and always check if the ID you chose is not already used by the core or one of the modules (see below for already existing IDs).
+
+| ID type           | Table               | Field      | Reserved IDs for the core | Reserved IDs for modules | Reserved IDs for custom use | Max value  |
+|-------------------|---------------------|------------|---------------------------|--------------------------|-----------------------------|------------|
+| Creature ID       | creature_template   | entry      | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 16777215   |
```
Comment (Kitzunu) [inline docs/Dealing-with-IDs.md]:
```suggestion
| Creature Entry       | [creature_template](creature_template#entry)   | entry      | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 16777215   |
```

File: docs/Dealing-with-IDs.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+| Creature GUID     | creature            | guid       | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 4294967295 |
```
Comment (Kitzunu) [inline docs/Dealing-with-IDs.md]:
```suggestion
| Creature GUID     | [creature](creature#guid)            | guid       | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 4294967295 |
```

File: docs/Dealing-with-IDs.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+| Gameobject ID     | gameobject_template | entry      | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 16777215   |
```
Comment (Kitzunu) [inline docs/Dealing-with-IDs.md]:
```suggestion
| Gameobject Entry     | [gameobject_template](gameobject_template#entry) | entry      | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 16777215   |
```

File: docs/Dealing-with-IDs.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+| Gameobject GUID   | gameobject          | guid       | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 4294967295 |
```
Comment (Kitzunu) [inline docs/Dealing-with-IDs.md]:
```suggestion
| Gameobject GUID   | [gameobject](gameobject#guid)          | guid       | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 4294967295 |
```

File: docs/Dealing-with-IDs.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+| NPC Text ID       | npc_text            | ID         | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 16777215   |
```
Comment (Kitzunu) [inline docs/Dealing-with-IDs.md]:
```suggestion
| NPC Text ID       | [npc_text](npc_text#id)            | ID         | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 16777215   |
```

File: docs/Dealing-with-IDs.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+| Gossip Menu ID    | gossip_menu         | MenuID     | < 57000                   | >= 57000 < 60000         | >= 60000                    | 65535      |
```
Comment (Kitzunu) [inline docs/Dealing-with-IDs.md]:
```suggestion
| Gossip Menu ID    | [gossip_menu](gossip_menu#menuid)         | MenuID     | < 57000                   | >= 57000 < 60000         | >= 60000                    | 65535      |
```

File: docs/Dealing-with-IDs.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+| Broadcast Text ID | broadcast_text      | ID         | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 16777215   |
```
Comment (Kitzunu) [inline docs/Dealing-with-IDs.md]:
```suggestion
| Broadcast Text ID | [broadcast_text](broadcast_text#id)      | ID         | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 16777215   |
```

File: docs/Dealing-with-IDs.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+| Creature Text ID  | creature_text       | CreatureID | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 16777215   |
```
Comment (Kitzunu) [inline docs/Dealing-with-IDs.md]:
```suggestion
| Creature Text ID  | [creature_text](creature_text#creatureid)       | CreatureID | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 16777215   |
```

File: docs/Dealing-with-IDs.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+| ACore String ID   | acore_string        | entry      | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 16777215   |
```
Comment (Kitzunu) [inline docs/Dealing-with-IDs.md]:
```suggestion
| ACore String Entry   | [acore_string](acore_string#entry)        | entry      | < 4000000                 | >= 4000000 < 5000000     | >= 5000000                  | 16777215   |
```

**Comment by Stoabrogga** [comment]
@Kitzunu 
Thanks for your feedback. I pushed a new commit removing the first column as the information is redundant. Instead I merged "Table / Field" and added links to the fields.

**Comment by Kitzunu** [comment]
Cool, great work :)🎉

**Comment by BarbzYHOOL** [comment]
@dunjeon this for custom IDs

**Comment by BarbzYHOOL** [comment]
we should make categories per table, not per arbitrary type like "creature", so i changed to "creature_template" as we can also have different tables for each creature ("creature_template_addon" etc)

**Comment by BarbzYHOOL** [comment]
ok i changed my mind, cause for simple `creature` we should not input the GUID, and for other tables, well ti's the same ID

**Comment by BarbzYHOOL** [comment]
ok let's merge, then more people can edit it

**Comment by BarbzYHOOL** [comment]
what about renaming the file to "custom_ids.md" ? because it's for custom IDs only

**Comment by BarbzYHOOL** [comment]
ok so i was looking at this file, the numbers have been arbitrarily decided and suggested which is fine
However, I think this should only be merged when we finally improve the current tables datatypes

Example: acore_string has currently mediumInt unsigned which can hold up to 16777215. But smallint unsigned can hold up to 65535 rows which is more than enough considering the biggest entry is 35410 actually (with tons of gaps too).
We could instead allocate a range like anything above 60000 = custom IDs.

So yeah, don't merge until we rework these

**Pull request closed without merging**